---
title: "05_downstream_analyses_pups"
author: "Anna Chamot"
date: '2025-06-27'
output: html_document
---

# ATAC-seq: Post-DAA functional follow-up analyses
This notebook contains downstream analyses following differential chromatin accessibility in PND15 spermatogonial stem cells, including peak annotation, motif and pathway enrichment, and integration with RNA-seq.

## 5.0 Setup and imports

```{r setup, include=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, verbose=FALSE}
# All required libraries for downstream annotation, plotting, pathway, motif analyses
suppressPackageStartupMessages({
  library(SummarizedExperiment)
  library(dplyr)
  library(ggplot2)
  library(cowplot)
  library(reshape2)
  library(EnhancedVolcano)
  library(pheatmap)
  library(dichromat)
  library(gprofiler2)
  library(limma)
  library(rtracklayer)
  library(ChIPseeker)
  library(TxDb.Mmusculus.UCSC.mm10.knownGene)
  library(org.Mm.eg.db)
  library(EnsDb.Mmusculus.v79)
  library(BSgenome.Mmusculus.UCSC.mm10)
  library(memes)
  library(tibble)
  library(dorothea)
})
```

## 5.1 Parameters, inputs, outputs
```{r, verbose=FALSE}
# Main directories and experimental settings (customize as needed)
outputdir <- "/mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/"
strain <- "B6"
generation <- "F1"
sex <- "Male"
Sub_age <- "P15"
suffix <- "Basemodel"
pthreshold <- 0.05
logthreshold <- 0.5

dirname <- paste0(strain, "-", generation, "-", sex, "-", Sub_age, "-", suffix, "/")
output <- paste0(outputdir, dirname)
if (!dir.exists(output)) dir.create(output, recursive = TRUE)
```

## 5.2 Load differential accessibility results
```{r, verbose=FALSE}
# Load SummarizedExperiment from previous DEA
SE <- readRDS(file = paste0(output, "SE_final.rds"))
```

## 5.3 Export Up/Downregulated peaks as BED
```{r, verbose=FALSE}
# Identify significantly up/down-regulated peaks
up <- SE[rowData(SE)$dea$padj < pthreshold & rowData(SE)$dea$logFC > logthreshold]
down <- SE[rowData(SE)$dea$padj < pthreshold & rowData(SE)$dea$logFC < -logthreshold]

# Extract GenomicRanges for up/down peaks
up.ranges <- granges(up)
down.ranges <- granges(down)

# Format BED: seqnames, start-1, end, strand
up.ranges.df <- data.frame(seqnames=seqnames(up.ranges), starts=start(up.ranges)-1, ends=end(up.ranges), strands=strand(up.ranges))
down.ranges.df <- data.frame(seqnames=seqnames(down.ranges), starts=start(down.ranges)-1, ends=end(down.ranges), strands=strand(down.ranges))

# Write BED files
write.table(up.ranges.df, file=paste0(output,"up.bed"), quote=F, sep="\t", row.names=F, col.names=F)
write.table(down.ranges.df, file=paste0(output,"down.bed"), quote=F, sep="\t", row.names=F, col.names=F)
```

## 5.4 Merge BAM files by condition (for downstream signal plotting)
```{r, verbose=FALSE}
# Sample lists by group
ctrBam <- colData(SE)$bamReads[colData(SE)$Condition == "Control"]
msusBam <- colData(SE)$bamReads[colData(SE)$Condition == "MSUS"]
```
```{r, eval=F, verbose=FALSE}
# Merge and index BAMs (system calls; check samtools availability)
system(paste0("samtools merge -@ 8 ", output, "ctrMerged.bam ", paste(ctrBam, collapse = " ")))
system(paste0("samtools index ", output, "ctrMerged.bam"))
system(paste0("samtools merge -@ 8 ", output, "msusMerged.bam ", paste(msusBam, collapse = " ")))
system(paste0("samtools index ", output, "msusMerged.bam"))
```

## 5.5 Generate signal tracks and heatmaps
```{bash engine.opts='-l', eval=F, verbose=FALSE}
# Merge BAM → bigWig (use your tracks.sh or other bigWig generator)
/mnt/groupMansuy/anna/Bioinformatics_project/tracks.sh /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/ctrMerged.bam /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/ /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/logs/41_mergedtracks/ 5 8

/mnt/groupMansuy/anna/Bioinformatics_project/tracks.sh /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/msusMerged.bam /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/ /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/logs/41_mergedtracks/ 5 8

# Compute matrix for up/down peaks and plot heatmap
computeMatrix reference-point -R /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/up.bed /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/down.bed --referencePoint "center" -S /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/ctrMerged.bw /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/msusMerged.bw -b 1000 -a 1000 --skipZeros -p 8 -o /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/DAR.mat.gz --outFileSortedRegions /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/DAR.bed

plotHeatmap -m /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/DAR.mat.gz --refPointLabel "Peak center" --plotFileFormat "pdf" --colorMap 'Greys' --whatToShow 'heatmap and colorbar' --heatmapHeight 10 -out /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/DAR.heatmap.pdf
```
<embed src="/mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/DAR.heatmap.pdf" width="800px" height="600px" type="application/pdf">

## 5.6 Motif Enrichment [AME + HOMER] 
```{bash engine.opts='-l', eval=F, verbose=FALSE}
# With HOMER
bed2pos.pl /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/up.bed > /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/up.pos

findMotifsGenome.pl /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/up.pos mm10 /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/up -preparsedDir /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/preparsed/ -size 200 -mask -p 8

bed2pos.pl /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/down.bed > /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/down.pos

findMotifsGenome.pl /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/down.pos mm10 /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/down -preparsedDir /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/preparsed/ -size 200 -mask -p 8
```

```{r, verbose=FALSE}
# Enriched motifs corresponding to the up-regulated peaks
homer_results_up <- read.delim(paste0(output,"up/knownResults.txt"), header = TRUE, stringsAsFactors = FALSE)
head(homer_results_up[, c("Motif.Name", "Consensus", "P.value", "q.value..Benjamini.")], 1)
```

```{r, verbose=FALSE}
# Enriched motifs corresponding to the down-regulated peaks
homer_results_down <- read.delim(paste0(output,"down/knownResults.txt"), header = TRUE, stringsAsFactors = FALSE)
head(homer_results_down[, c("Motif.Name", "Consensus", "P.value", "q.value..Benjamini.")], 5)
```

```{r, verbose=FALSE}
# With MEME/AME tools
# Extract sequences from the genome (use the appropriate genome, here mouse mm10)
down.sequences <- getSeq(BSgenome.Mmusculus.UCSC.mm10, down.ranges)
up.sequences <- getSeq(BSgenome.Mmusculus.UCSC.mm10, up.ranges)

# Save the sequences in FASTA format for use with AME
Biostrings::writeXStringSet(down.sequences, filepath = paste0(output, "down_sequences.fasta"))
Biostrings::writeXStringSet(up.sequences, filepath = paste0(output, "up_sequences.fasta"))

# Run AME with the input sequences and background sequences
ame_results <- runAme(input = paste0(output, "down_sequences.fasta"), control = "shuffle",     
  database = "/mnt/groupMansuy/leo/references/JASPAR/JASPAR2024_CORE.meme", outdir = paste0(output, "ame_output_down"), meme_path = "/mnt/IM/meme/bin/", evalue_report_threshold = 0.05)

ame_results <- runAme(input = paste0(output, "up_sequences.fasta"), control = "shuffle", database = "/mnt/groupMansuy/leo/references/JASPAR/JASPAR2024_CORE.meme", outdir = paste0(output, "ame_output_up"), meme_path = "/mnt/IM/meme/bin/", evalue_report_threshold = 0.05)
```

```{r, verbose=FALSE}
# Read the AME results .tsv file, for up-regulated peaks
ame_results <- read.delim(paste0(output, "ame_output_up/ame.tsv"), header = TRUE)
print(dim(ame_results))
```

```{r, verbose=FALSE}
# Read the AME results .tsv file, for down-regulated peaks
ame_results <- read.delim(paste0(output, "ame_output_down/ame.tsv"), header = TRUE)
head(ame_results[, c("motif_alt_ID", "consensus", "p.value", "adj_p.value")], 5)
```

## 5.7 Correlate ATAC-seq and RNA-seq results
```{r, verbose=FALSE, message=FALSE}
# Load and annotate peaks with nearest gene using ChIPseeker
atac.res.ranges <- granges(SE)
txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
peaks.anno.atac.res <- annotatePeak(atac.res.ranges, tssRegion = c(-3000, 3000), TxDb = txdb, annoDb = "org.Mm.eg.db")
df.atac <- as.data.frame(peaks.anno.atac.res@anno)
df.atac.prom <- subset(df.atac, grepl("Promoter", annotation))

# Load RNA-seq SummarizedExperiment
SE.RNA <- readRDS("/mnt/IM/groupMansuy/theresa/RNAseq/MSUS39F1_SC/results_mm10/03_dea/results/PND15_Basemodel/SE.rds")
res.rna <- as.data.frame(rowData(SE.RNA))
res.rna <- rownames_to_column(res.rna, var = "ENSEMBL")

# Join ATAC promoter peaks to RNA-seq results by gene ENSEMBL ID
df.combined <- left_join(df.atac.prom, res.rna, by = "ENSEMBL")
df.combined <- df.combined[!is.na(df.combined$dea.logFC.y),]
```
```{r, verbose=FALSE}
# Correlation: all promoters
plot(df.combined$dea.logFC.x, df.combined$dea.logFC.y, xlab="ATAC logFC", ylab="RNA logFC")
cor.test(df.combined$dea.logFC.x, df.combined$dea.logFC.y)

# Significant ATAC, what proportion also significant in RNA?
df.combined.sign.atac <- df.combined[df.combined$dea.padj.x < pthreshold,]
sum(df.combined.sign.atac$dea.padj.y < pthreshold) / nrow(df.combined.sign.atac)

# Significant RNA, what proportion also significant in ATAC?
df.combined.sign.rna <- df.combined[df.combined$dea.padj.y < pthreshold,]
sum(df.combined.sign.rna$dea.padj.x < pthreshold) / nrow(df.combined.sign.rna)

# Both significant
df.combined.sign.both <- df.combined[df.combined$dea.padj.x < pthreshold & df.combined$dea.padj.y < pthreshold,]
dim(df.combined.sign.both)
```

## 5.8 Export differential genes for heatmaps (for deepTools etc.)
```{r, verbose=FALSE, message=FALSE}
# Get differentially expressed genes (padj and/or PValue)
res <- res.rna
colnames(res) <- gsub("^dea\\.", "", colnames(res))

# By adjusted p-value
res.up <- res[res$padj < pthreshold & res$logFC > logthreshold,]
res.down <- res[res$padj < pthreshold & res$logFC < -logthreshold,]

entrez_ids_up <- mapIds(org.Mm.eg.db, keys = res.up$ENSEMBL, column = "ENTREZID", keytype = "ENSEMBL", multiVals = "first")
entrez_ids_down <- mapIds(org.Mm.eg.db, keys = res.down$ENSEMBL, column = "ENTREZID", keytype = "ENSEMBL", multiVals = "first")

# Whole gene and gene start BEDs
genes <- genes(TxDb.Mmusculus.UCSC.mm10.knownGene)
genes.up <- genes[genes$gene_id %in% entrez_ids_up]
genes.down <- genes[genes$gene_id %in% entrez_ids_down]
export.bed(genes.up, con = paste0(output, "genes.up.bed"))
export.bed(genes.down, con = paste0(output, "genes.down.bed"))

genes.start.up <- genes.up; end(genes.start.up) <- start(genes.start.up)
genes.start.down <- genes.down; end(genes.start.down) <- start(genes.start.down)
export.bed(genes.start.up, con = paste0(output, "genes.start.up.bed"))
export.bed(genes.start.down, con = paste0(output, "genes.start.down.bed"))
```

## 5.9 Pathway analysis of differentially accessible genes
```{r, verbose=FALSE, message=FALSE}
# Annotate peak-associated promoters for GO enrichment (using g:Profiler)
peaks.anno.up <- annotatePeak(readPeakFile(paste0(output, "up.bed")), tssRegion = c(-3000, 3000), TxDb = txdb, annoDb = "org.Mm.eg.db")
peaks.anno.down <- annotatePeak(readPeakFile(paste0(output, "down.bed")), tssRegion = c(-3000, 3000), TxDb = txdb, annoDb = "org.Mm.eg.db")
```
```{r, verbose=FALSE}
promoter_peaks.up <- subset(as.data.frame(peaks.anno.up), grepl("Promoter", annotation))
promoter_peaks.down <- subset(as.data.frame(peaks.anno.down), grepl("Promoter", annotation))
promoter_peaks.merged <- rbind(promoter_peaks.up, promoter_peaks.down)

go_up <- gost(query = promoter_peaks.up$ENSEMBL, organism = "mmusculus", ordered_query = FALSE, multi_query = FALSE, significant = TRUE, exclude_iea = FALSE, measure_underrepresentation = FALSE, evcodes = TRUE, user_threshold = 0.05, correction_method = "g_SCS", domain_scope = "annotated", custom_bg = NULL, numeric_ns = "", sources = "GO:BP", as_short_link = FALSE, highlight = TRUE)

go_down <- gost(query = promoter_peaks.down$ENSEMBL, organism = "mmusculus", ordered_query = FALSE, multi_query = FALSE, significant = TRUE, exclude_iea = FALSE, measure_underrepresentation = FALSE, evcodes = TRUE, user_threshold = 0.05, correction_method = "g_SCS", domain_scope = "annotated", custom_bg = NULL, numeric_ns = "", sources = "GO:BP", as_short_link = FALSE, highlight = TRUE)

go_merged <- gost(query = promoter_peaks.merged$ENSEMBL, organism = "mmusculus", ordered_query = FALSE, multi_query = FALSE, significant = TRUE, exclude_iea = FALSE, measure_underrepresentation = FALSE, evcodes = TRUE, user_threshold = 0.05, correction_method = "g_SCS", domain_scope = "annotated", custom_bg = NULL, numeric_ns = "", sources = "GO:BP", as_short_link = FALSE, highlight = TRUE)
```

## 5.10 Motif Enrichment: TFs & Target Regulation in RNA-seq
```{r, verbose=FALSE, message=FALSE}
# Use dorothea for TF target annotation
dorothea <- dorothea_mm[dorothea_mm$confidence %in% c("A", "B", "C", "D", "E"),]

# Example: load top TF motifs from AME output for downregulated peaks
tf.down <- read.delim(paste0(output, "ame_output_down/ame.tsv"))
topTFs <- 5
summarycols <- c("TF", "TargetsTotal", "TargetsInRNA", "TargetsPValueSign", "TargetsPadjSign", "Up", "Down", "NoChange", "PercUp", "PercDown", "PercNoChange")
summary.down <- as.data.frame(matrix(NA, nrow = topTFs, ncol = length(summarycols), dimnames = list(NULL, summarycols)))

for (i in 1:topTFs) {
  pos <- grep(tf.down$motif_alt_ID[i], dorothea$tf, ignore.case = TRUE)
  targets_symbols <- dorothea$target[pos]
  targets <- mapIds(org.Mm.eg.db, keys = targets_symbols, column = "ENSEMBL", keytype = "SYMBOL", multiVals = "first")
  SE.RNA.sub <- SE.RNA[rownames(SE.RNA) %in% targets,]
  res.sub <- rowData(SE.RNA.sub)$dea
  summary.down[i,] <- list(tf.down$motif_alt_ID[i], length(pos), nrow(SE.RNA.sub), sum(res.sub$PValue < pthreshold), sum(res.sub$padj < pthreshold), sum(res.sub$logFC > logthreshold), sum(res.sub$logFC < -logthreshold), nrow(SE.RNA.sub) - sum(res.sub$logFC > logthreshold) - sum(res.sub$logFC < -logthreshold), sum(res.sub$logFC > logthreshold)/nrow(SE.RNA.sub), sum(res.sub$logFC < -logthreshold)/nrow(SE.RNA.sub), (nrow(SE.RNA.sub) - sum(res.sub$logFC > logthreshold) - sum(res.sub$logFC < -logthreshold))/nrow(SE.RNA.sub))
}
# Visualization: stacked barplot (Up/Down/NoChange) per TF
summary.down$TF <- factor(summary.down$TF, levels = summary.down$TF[order(summary.down$PercDown)])
summary.down.melted <- reshape2::melt(summary.down, id.vars = "TF", measure.vars = c("PercUp", "PercDown", "PercNoChange"))
ggplot(summary.down.melted, aes(x = TF, y = value, fill = variable)) + 
  geom_bar(stat = "identity") + scale_fill_manual(values = c("PercUp" = "blue", "PercDown" = "red", "PercNoChange" = "grey"), labels = c("Upregulated", "Downregulated", "No changes")) + labs(x = "TF", y = "Fraction of targets", fill = "Regulation") + ggtitle("TF motif enrichment: Down peaks, target gene regulation in RNA-seq") + theme_minimal() + theme(axis.text.x = element_text(angle = 60, hjust = 1))
```

# Analysis begin less stringent with thresholds (Pvalue < 0.05 here)

## 5.1.2 Parameters, inputs, outputs
```{r, verbose=FALSE}
# Main directories and experimental settings (customize as needed)
outputdir <- "/mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/"
strain <- "B6"
generation <- "F1"
sex <- "Male"
Sub_age <- "P15"
suffix <- "Basemodel"
pthreshold <- 0.05
logthreshold <- 0.5

dirname <- paste0(strain, "-", generation, "-", sex, "-", Sub_age, "-", suffix, "/")
output <- paste0(outputdir, dirname)
output <- paste0(output, "Pval_inf_005/")
if (!dir.exists(output)) dir.create(output, recursive = TRUE)
```

## 5.2.2 Load differential accessibility results
```{r, verbose=FALSE}
# Load SummarizedExperiment from previous DEA
SE <- readRDS(file = paste0("/mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/", "SE_final.rds"))
```

## 5.3.2 Export Up/Downregulated peaks as BED
```{r, verbose=FALSE}
# Identify significantly up/down-regulated peaks
up <- SE[rowData(SE)$dea$PValue < pthreshold & rowData(SE)$dea$logFC > logthreshold]
down <- SE[rowData(SE)$dea$PValue < pthreshold & rowData(SE)$dea$logFC < -logthreshold]

# Extract GenomicRanges for up/down peaks
up.ranges <- granges(up)
down.ranges <- granges(down)

# Format BED: seqnames, start-1, end, strand
up.ranges.df <- data.frame(seqnames=seqnames(up.ranges), starts=start(up.ranges)-1, ends=end(up.ranges), strands=strand(up.ranges))
down.ranges.df <- data.frame(seqnames=seqnames(down.ranges), starts=start(down.ranges)-1, ends=end(down.ranges), strands=strand(down.ranges))

# Write BED files
write.table(up.ranges.df, file=paste0(output,"up.bed"), quote=F, sep="\t", row.names=F, col.names=F)
write.table(down.ranges.df, file=paste0(output,"down.bed"), quote=F, sep="\t", row.names=F, col.names=F)
```

## 5.4.2 Merge BAM files by condition (for downstream signal plotting)
```{r, verbose=FALSE}
# Sample lists by group
ctrBam <- colData(SE)$bamReads[colData(SE)$Condition == "Control"]
msusBam <- colData(SE)$bamReads[colData(SE)$Condition == "MSUS"]
```
```{r, eval=F, verbose=FALSE}
# Merge and index BAMs (system calls; check samtools availability)
system(paste0("samtools merge -@ 8 ", output, "ctrMerged.bam ", paste(ctrBam, collapse = " ")))
system(paste0("samtools index ", output, "ctrMerged.bam"))
system(paste0("samtools merge -@ 8 ", output, "msusMerged.bam ", paste(msusBam, collapse = " ")))
system(paste0("samtools index ", output, "msusMerged.bam"))
```

## 5.5.2 Generate signal tracks and heatmaps
```{bash engine.opts='-l', eval=F, verbose=FALSE}
# Merge BAM → bigWig (use your tracks.sh or other bigWig generator)
/mnt/groupMansuy/anna/Bioinformatics_project/tracks.sh /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/ctrMerged.bam /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/ /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/logs/41_mergedtracks/ 5 8

/mnt/groupMansuy/anna/Bioinformatics_project/tracks.sh /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/msusMerged.bam /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/ /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/logs/41_mergedtracks/ 5 8

# Compute matrix for up/down peaks and plot heatmap
computeMatrix reference-point -R /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/up.bed /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/down.bed --referencePoint "center" -S /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/ctrMerged.bw /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/msusMerged.bw -b 1000 -a 1000 --skipZeros -p 8 -o /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/DAR.mat.gz --outFileSortedRegions /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/DAR.bed

plotHeatmap -m /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/DAR.mat.gz --refPointLabel "Peak center" --plotFileFormat "pdf" --colorMap 'Greys' --whatToShow 'heatmap and colorbar' --heatmapHeight 10 -out /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/DAR.heatmap.pdf
```
<embed src="/mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/DAR.heatmap.pdf" width="800px" height="600px" type="application/pdf">

## 5.6.2 Motif Enrichment [AME + HOMER] 
```{bash engine.opts='-l', eval=F, verbose=FALSE}
# With HOMER
bed2pos.pl /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/up.bed > /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/up.pos

findMotifsGenome.pl /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/up.pos mm10 /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/up -preparsedDir /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/preparsed/ -size 200 -mask -p 8

bed2pos.pl /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/down.bed > /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/down.pos

findMotifsGenome.pl /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/down.pos mm10 /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/down -preparsedDir /mnt/groupMansuy/anna/Bioinformatics_project/ATAC/results/dea_04/B6-F1-Male-P15-Basemodel/Pval_inf_005/preparsed/ -size 200 -mask -p 8
```

```{r, verbose=FALSE}
# Enriched motifs corresponding to the up-regulated peaks
homer_results_up <- read.delim(paste0(output,"up/knownResults.txt"), header = TRUE, stringsAsFactors = FALSE)
head(homer_results_up[, c("Motif.Name", "Consensus", "P.value", "q.value..Benjamini.")], 10)
```

```{r, verbose=FALSE}
# Enriched motifs corresponding to the down-regulated peaks
homer_results_down <- read.delim(paste0(output,"down/knownResults.txt"), header = TRUE, stringsAsFactors = FALSE)
head(homer_results_down[, c("Motif.Name", "Consensus", "P.value", "q.value..Benjamini.")], 10)
```

```{r, verbose=FALSE}
# With MEME/AME tools
# Extract sequences from the genome (use the appropriate genome, here mouse mm10)
down.sequences <- getSeq(BSgenome.Mmusculus.UCSC.mm10, down.ranges)
up.sequences <- getSeq(BSgenome.Mmusculus.UCSC.mm10, up.ranges)

# Save the sequences in FASTA format for use with AME
Biostrings::writeXStringSet(down.sequences, filepath = paste0(output, "down_sequences.fasta"))
Biostrings::writeXStringSet(up.sequences, filepath = paste0(output, "up_sequences.fasta"))

# Run AME with the input sequences and background sequences
ame_results <- runAme(input = paste0(output, "down_sequences.fasta"), control = "shuffle", database = "/mnt/groupMansuy/leo/references/JASPAR/JASPAR2024_CORE.meme", outdir = paste0(output, "ame_output_down"), meme_path = "/mnt/IM/meme/bin/", evalue_report_threshold = 0.05)

ame_results <- runAme(input = paste0(output, "up_sequences.fasta"), control = "shuffle", database = "/mnt/groupMansuy/leo/references/JASPAR/JASPAR2024_CORE.meme", outdir = paste0(output, "ame_output_up"), meme_path = "/mnt/IM/meme/bin/", evalue_report_threshold = 0.05)
```

```{r, verbose=FALSE}
# Read the AME results .tsv file, for up-regulated peaks
ame_results <- read.delim(paste0(output, "ame_output_up/ame.tsv"), header = TRUE)
head(ame_results[, c("motif_alt_ID", "consensus", "p.value", "adj_p.value")], 10)
```

```{r, verbose=FALSE}
# Read the AME results .tsv file, for down-regulated peaks
ame_results <- read.delim(paste0(output, "ame_output_down/ame.tsv"), header = TRUE)
head(ame_results[, c("motif_alt_ID", "consensus", "p.value", "adj_p.value")], 10)
```

## 5.7.2 Correlate ATAC-seq and RNA-seq results
```{r, verbose=FALSE, message=FALSE}
# Load and annotate peaks with nearest gene using ChIPseeker
atac.res.ranges <- granges(SE)
txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
peaks.anno.atac.res <- annotatePeak(atac.res.ranges, tssRegion = c(-3000, 3000), TxDb = txdb, annoDb = "org.Mm.eg.db")
df.atac <- as.data.frame(peaks.anno.atac.res@anno)
df.atac.prom <- subset(df.atac, grepl("Promoter", annotation))

# Load RNA-seq SummarizedExperiment
SE.RNA <- readRDS("/mnt/IM/groupMansuy/theresa/RNAseq/MSUS39F1_SC/results_mm10/03_dea/results/PND15_Basemodel/SE.rds")
res.rna <- as.data.frame(rowData(SE.RNA))
res.rna <- rownames_to_column(res.rna, var = "ENSEMBL")

# Join ATAC promoter peaks to RNA-seq results by gene ENSEMBL ID
df.combined <- left_join(df.atac.prom, res.rna, by = "ENSEMBL")
df.combined <- df.combined[!is.na(df.combined$dea.logFC.y),]
```
```{r, verbose=FALSE}
# Correlation: all promoters
plot(df.combined$dea.logFC.x, df.combined$dea.logFC.y, xlab="ATAC logFC", ylab="RNA logFC")
cor.test(df.combined$dea.logFC.x, df.combined$dea.logFC.y)

# Significant ATAC, what proportion also significant in RNA?
df.combined.sign.atac <- df.combined[df.combined$dea.padj.x < pthreshold,]
sum(df.combined.sign.atac$dea.padj.y < pthreshold) / nrow(df.combined.sign.atac)

# Significant RNA, what proportion also significant in ATAC?
df.combined.sign.rna <- df.combined[df.combined$dea.padj.y < pthreshold,]
sum(df.combined.sign.rna$dea.padj.x < pthreshold) / nrow(df.combined.sign.rna)

# Both significant
df.combined.sign.both <- df.combined[df.combined$dea.padj.x < pthreshold & df.combined$dea.padj.y < pthreshold,]
dim(df.combined.sign.both)
```

## 5.8.2 Export differential genes for heatmaps (for deepTools etc.)
```{r, verbose=FALSE, message=FALSE}
# Whole gene and gene start BEDs
export.bed(genes.up, con = paste0(output, "genes.up.bed"))
export.bed(genes.down, con = paste0(output, "genes.down.bed"))

export.bed(genes.start.up, con = paste0(output, "genes.start.up.bed"))
export.bed(genes.start.down, con = paste0(output, "genes.start.down.bed"))
```

# 5.9.2 Pathway analysis of differentially accessible genes
```{r, verbose=FALSE, message=FALSE}
# Annotate peak-associated promoters for GO enrichment (using g:Profiler)
peaks.anno.up <- annotatePeak(readPeakFile(paste0(output, "up.bed")), tssRegion = c(-3000, 3000), TxDb = txdb, annoDb = "org.Mm.eg.db")
peaks.anno.down <- annotatePeak(readPeakFile(paste0(output, "down.bed")), tssRegion = c(-3000, 3000), TxDb = txdb, annoDb = "org.Mm.eg.db")
```
```{r, verbose=FALSE}
promoter_peaks.up <- subset(as.data.frame(peaks.anno.up), grepl("Promoter", annotation))
promoter_peaks.down <- subset(as.data.frame(peaks.anno.down), grepl("Promoter", annotation))
promoter_peaks.merged <- rbind(promoter_peaks.up, promoter_peaks.down)

go_up <- gost(query = promoter_peaks.up$ENSEMBL, organism = "mmusculus", ordered_query = FALSE, multi_query = FALSE, significant = TRUE, exclude_iea = FALSE, measure_underrepresentation = FALSE, evcodes = TRUE, user_threshold = 0.05, correction_method = "g_SCS", domain_scope = "annotated", custom_bg = NULL, numeric_ns = "", sources = "GO:BP", as_short_link = FALSE, highlight = TRUE)
go_down <- gost(query = promoter_peaks.down$ENSEMBL, organism = "mmusculus", ordered_query = FALSE, multi_query = FALSE, significant = TRUE, exclude_iea = FALSE, measure_underrepresentation = FALSE, evcodes = TRUE, user_threshold = 0.05, correction_method = "g_SCS", domain_scope = "annotated", custom_bg = NULL, numeric_ns = "", sources = "GO:BP", as_short_link = FALSE, highlight = TRUE)
go_merged <- gost(query = promoter_peaks.merged$ENSEMBL, organism = "mmusculus", ordered_query = FALSE, multi_query = FALSE, significant = TRUE, exclude_iea = FALSE, measure_underrepresentation = FALSE, evcodes = TRUE, user_threshold = 0.05, correction_method = "g_SCS", domain_scope = "annotated", custom_bg = NULL, numeric_ns = "", sources = "GO:BP", as_short_link = FALSE, highlight = TRUE)
```
```{r, verbose=FALSE}
head(go_up$result[, c("p_value", "term_size", "query_size", "intersection_size", "term_name")])
head(go_down$result[, c("p_value", "term_size", "query_size", "intersection_size", "term_name")])
head(go_merged$result[, c("p_value", "term_size", "query_size", "intersection_size", "term_name")], 10)
```

# 5.10.2 Motif Enrichment: TFs & Target Regulation in RNA-seq
```{r, verbose=FALSE, message=FALSE}
# Use dorothea for TF target annotation
dorothea <- dorothea_mm[dorothea_mm$confidence %in% c("A", "B", "C", "D", "E"),]

# Example: load top TF motifs from AME output for downregulated peaks
tf.down <- read.delim(paste0(output, "ame_output_down/ame.tsv"))
topTFs <- 50
summarycols <- c("TF", "TargetsTotal", "TargetsInRNA", "TargetsPValueSign", "TargetsPadjSign", "Up", "Down", "NoChange", 
                 "PercUp", "PercDown", "PercNoChange")
summary.down <- as.data.frame(matrix(NA, nrow = topTFs, ncol = length(summarycols), dimnames = list(NULL, summarycols)))

for (i in 1:topTFs) {
  pos <- grep(tf.down$motif_alt_ID[i], dorothea$tf, ignore.case = TRUE)
  # Skip if no targets found
  if (length(pos) == 0) {
    message("No dorothea targets found for motif: ", tf.down$motif_alt_ID[i])
    next
  }
  targets_symbols <- dorothea$target[pos]
  # Skip if no target symbols
  if (length(targets_symbols) == 0) {
    message("No gene symbols for motif: ", tf.down$motif_alt_ID[i])
    next
  }
  targets <- mapIds(org.Mm.eg.db, keys = targets_symbols, column = "ENSEMBL", keytype = "SYMBOL", multiVals = "first")
  # Remove any NAs from targets (no mapping)
  targets <- targets[!is.na(targets)]
  # Skip if no ENSEMBL IDs found
  if (length(targets) == 0) {
    message("No ENSEMBL IDs for motif: ", tf.down$motif_alt_ID[i])
    next
  }
  SE.RNA.sub <- SE.RNA[rownames(SE.RNA) %in% targets,]
  # If no overlap in RNA data, skip
  if (nrow(SE.RNA.sub) == 0) {
    message("No RNA-seq data for motif: ", tf.down$motif_alt_ID[i])
    next
  }
  res.sub <- rowData(SE.RNA.sub)$dea
  summary.down[i,] <- list(
    tf.down$motif_alt_ID[i], length(pos), nrow(SE.RNA.sub),
    sum(res.sub$PValue < pthreshold), sum(res.sub$padj < pthreshold),
    sum(res.sub$logFC > logthreshold), sum(res.sub$logFC < -logthreshold),
    nrow(SE.RNA.sub) - sum(res.sub$logFC > logthreshold) - sum(res.sub$logFC < -logthreshold),
    sum(res.sub$logFC > logthreshold)/nrow(SE.RNA.sub),
    sum(res.sub$logFC < -logthreshold)/nrow(SE.RNA.sub),
    (nrow(SE.RNA.sub) - sum(res.sub$logFC > logthreshold) - sum(res.sub$logFC < -logthreshold))/nrow(SE.RNA.sub)
  )
}
# Visualization: stacked barplot (Up/Down/NoChange) per TF
summary.down$TF <- factor(summary.down$TF, levels = summary.down$TF[order(summary.down$PercDown)])
summary.down.melted <- reshape2::melt(summary.down, id.vars = "TF", measure.vars = c("PercUp", "PercDown", "PercNoChange"))
ggplot(summary.down.melted, aes(x = TF, y = value, fill = variable)) + 
  geom_bar(stat = "identity") +
  scale_fill_manual(values = c("PercUp" = "blue", "PercDown" = "red", "PercNoChange" = "grey"),
                    labels = c("Upregulated", "Downregulated", "No changes")) +
  labs(x = "TF", y = "Fraction of targets", fill = "Regulation") +
  ggtitle("TF motif enrichment: Down peaks, target gene regulation in RNA-seq") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 60, hjust = 1))
```

## 5.11 Interpretation
This integrative analysis of chromatin accessibility (ATAC-seq) and gene expression (RNA-seq) in spermatogonial stem cells at PND15 following early-life stress (MSUS) reveals that stress induces widespread changes in chromatin accessibility. Both upregulated and downregulated regions were identified, with the most robust and statistically significant motif enrichment observed among downregulated peaks, specifically for STAT family transcription factor motifs, when peaks were selected using stringent, FDR-adjusted thresholds. This indicates that early-life stress is associated with decreased accessibility at key regulatory sites linked to immune and cytokine signaling pathways.  
For upregulated regions, no motifs remained significant after correction for multiple testing when using FDR-adjusted thresholds; however, enrichment for testis-specific and AP-1 family motifs was observed when a relaxed (unadjusted p-value) threshold was applied. This highlights that some biologically relevant regulatory signals may only become detectable with broader peak selection criteria.  
Pathway analysis further demonstrated that associations with developmental and differentiation processes are primarily seen when differentially accessible regions are defined using these relaxed thresholds, rather than more stringent criteria. Importantly, despite extensive chromatin remodeling, the correlation between chromatin accessibility changes and gene expression was minimal, and there was almost no overlap in significantly affected genes between ATAC-seq and RNA-seq datasets.  
Overall, these findings indicate that early-life stress leads to significant, motif- and pathway-specific changes in the epigenomic landscape of spermatogonial stem cells, most notably a reduction in accessibility at STAT-regulated loci. However, these chromatin changes are largely uncoupled from immediate transcriptional changes, suggesting a primed but not yet fully realized regulatory state. This poised chromatin configuration may underlie long-term functional consequences for germ cell development and future gene regulation.
